blockchain/utxocache.go

```diff
@@ -126,3 +133,410 @@ func (ms *mapSlice) deleteMaps() {
+	// cached entries.
+	if entry.isFresh() {
+		// If the entry is fresh, we will always have it in the cache.
+		s.cachedEntries.delete(txIn.PreviousOutPoint)
+		s.totalEntryMemory -= entry.memoryUsage()
+	} else {
+		// Can leave the entry to be garbage collected as the only purpose
+		// of this entry now is so that the entry on disk can be deleted.
+		entry = nil
+		s.totalEntryMemory -= entry.memoryUsage()
+	}
+
+	return nil
+}
+
+// addTxIns will add the given inputs of the tx if it's not a coinbase tx and if
+// the previous output that the input is pointing to exists in the utxo set.  The
+// utxo that is being spent by the input will be marked as spent and if the utxo
+// is fresh (meaning that the database on disk never saw it), it will be removed
+// from the cache.
+func (s *utxoCache) addTxIns(tx *btcutil.Tx, stxos *[]SpentTxOut) error {
+	// Coinbase transactions don't have any inputs to spend.
+	if IsCoinBase(tx) {
+		return nil
+	}
+
+	for _, txIn := range tx.MsgTx().TxIn {
+		err := s.addTxIn(txIn, stxos)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// connectTransaction updates the cache by adding all new utxos created by the
+// passed transaction and marking and/or removing all utxos that the transactions
+// spend as spent.  In addition, when the 'stxos' argument is not nil, it will
+// be updated to append an entry for each spent txout.  An error will be returned
+// if the cache and the database does not contain the required utxos.
+func (s *utxoCache) connectTransaction(
+	tx *btcutil.Tx, blockHeight int32, stxos *[]SpentTxOut) error {
+
+	err := s.addTxIns(tx, stxos)
+	if err != nil {
+		return err
+	}
+
+	// Add the transaction's outputs as available utxos.
+	return s.addTxOuts(tx, blockHeight)
+}
+
+// connectTransactions updates the cache by adding all new utxos created by all
+// of the transactions in the passed block, marking and/or removing all utxos
+// the transactions spend as spent, and setting the best hash for the view to
+// the passed block.  In addition, when the 'stxos' argument is not nil, it will
+// be updated to append an entry for each spent txout.
+func (s *utxoCache) connectTransactions(block *btcutil.Block, stxos *[]SpentTxOut) error {
+	for _, tx := range block.Transactions() {
+		err := s.connectTransaction(tx, block.Height(), stxos)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// writeCache writes all the entries that are cached in memory to the database atomically.
+func (s *utxoCache) writeCache(bestState *BestState) error {
+	// Update commits and flushes the cache to the database.
+	// NOTE: The database has its own cache which gets atomically written
+	// to leveldb.
+	err := s.db.Update(func(dbTx database.Tx) error {
+		utxoBucket := dbTx.Metadata().Bucket(utxoSetBucketName)
+		for i := range s.cachedEntries.maps {
+			for outpoint, entry := range s.cachedEntries.maps[i] {
+				// If the entry is nil or spent, remove the entry from the database
```

Comment:
Style nit: `delete(s.cachedEntries.maps[i], outpoint)` appears 3 times in the loop, we can reduce it to one: 
```go
switch {
    case  entry == nil || entry.IsSpent():
        	err := dbDeleteUtxoEntry(utxoBucket, outpoint)
		if err != nil {
			return err
		}

    case !entry.isModified():
   default:
		// Entry is fresh and needs to be put into the database.
		err := dbPutUtxoEntry(utxoBucket, outpoint, entry)
		if err != nil {
			return err
		}
}

delete(s.cachedEntries.maps[i], outpoint)
```

Reply:
Addressed